### PR TITLE
feat: queue notarize request

### PR DESCRIPTION
**Pull request by tl240**
This PR add an AWS SQS to queue the notarization requests to TaskFi-TLSN proxy server. Instead of calling `/notarize` API directly, now the notarization requests are queued in SQS for the consumers to pull and process in order.

For local testing, a [LocalStack](https://www.localstack.cloud/) docker container is set up in the project. To start docker containers for local testing, please run:

```
docker compose --profile dev up -d
```

Please refer to [PR](https://github.com/icetea-labs/taskfi-worker/pull/55) for the changes in TaskFi worker repository.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -13,6 +16,13 @@ mod routes;
 mod startup;
 mod telemetry;
 
+const MAX_THREADS_LIMIT: usize = 20;
```
Comment (tino1304) [inline src/main.rs]: đưa vô env

File: docker-init/localstack/sqs.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+LOCALSTACK_HOST=localhost
+AWS_REGION=ap-southeast-1
+
+create_queue() {
+    local QUEUE_NAME_TO_CREATE=$1
+    awslocal --endpoint-url=http://${LOCALSTACK_HOST}:4566 sqs create-queue \
+        --queue-name ${QUEUE_NAME_TO_CREATE} \
+        --region ${AWS_REGION} \
+        --attributes VisibilityTimeout=30
```
Comment (tino1304) [inline docker-init/localstack/sqs.sh]: - 30s khả năng ko đủ để process trên production
- có cần handle duplication?
- https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/standard-queues-at-least-once-delivery.html

File: src/main.rs [hunk 3]
```diff
@@ -32,6 +42,21 @@ async fn main() -> std::io::Result<()> {
...
+    let num_threads = std::thread::available_parallelism().unwrap().get();
+    let rt = runtime::Builder::new_multi_thread()
+        .worker_threads(min(num_threads, MAX_THREADS_LIMIT))
+        .enable_all()
+        .build()
+        .unwrap();
+
+    // Using a semaphore to ensure only 1 worker can pull messages from queue at
+    // a time, thus avoid the case where the same message may be consumed twice
+    let semaphore = Arc::new(Semaphore::new(1));
```
Comment (tino1304) [inline src/main.rs]: sqs visiblity timeout có thể handle được case này rồi?

File: docker-init/localstack/sqs.sh (hunk 2 above)
Comment (tino1304) [inline docker-init/localstack/sqs.sh]: 30s liệu có đủ trên production ko?

File: src/main.rs (hunk 1 above)
Comment (tino1304) [inline src/main.rs]: đưa vào env

File: src/main.rs [hunk 4]
```diff
@@ -32,6 +42,21 @@ async fn main() -> std::io::Result<()> {
...
     println!("Starting server at http://{}", app_address);
 
+    let num_threads = std::thread::available_parallelism().unwrap().get();
+    let rt = runtime::Builder::new_multi_thread()
+        .worker_threads(min(num_threads, MAX_THREADS_LIMIT))
+        .enable_all()
+        .build()
+        .unwrap();
+
+    // Using a semaphore to ensure only 1 worker can pull messages from queue at
```
Comment (tino1304) [inline src/main.rs]: handle duplication chưa chuẩn: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/standard-queues-at-least-once-delivery.html

File: src/main.rs (hunk 1 above)
Comment (tl240) [reply src/main.rs]: Em đã sửa trong commit [108a83c](https://github.com/MintIcetea/taskfi-tlsn/pull/4/commits/108a83c30ff191a6a93b55c07e0bf1e1bb971ba1)

File: docker-init/localstack/sqs.sh (hunk 2 above)
Comment (tl240) [reply docker-init/localstack/sqs.sh]: - Em đã tăng lên 120s trong commit [108a83c](https://github.com/MintIcetea/taskfi-tlsn/pull/4/commits/108a83c30ff191a6a93b55c07e0bf1e1bb971ba1)
- Image này chỉ để giả lập lại SQS trên môi trường development, khi chạy trên production sẽ dùng AWS SQS nên config này không ảnh hưởng đến môi trường production.

File: src/main.rs (hunk 3 above)
Comment (tl240) [reply src/main.rs]: Em đã sửa trong commit [108a83c](https://github.com/MintIcetea/taskfi-tlsn/pull/4/commits/108a83c30ff191a6a93b55c07e0bf1e1bb971ba1)

File: src/main.rs (hunk 4 above)
Comment (tl240) [reply src/main.rs]: Em đã sửa trong commit [108a83c](https://github.com/MintIcetea/taskfi-tlsn/pull/4/commits/108a83c30ff191a6a93b55c07e0bf1e1bb971ba1), đổi sang check duplication bằng map `request_id`

File: docker-init/localstack/sqs.sh (hunk 2 above)
Comment (tl240) [reply docker-init/localstack/sqs.sh]: Duplicate của comment trên, em đã sửa trong commit [108a83c](https://github.com/MintIcetea/taskfi-tlsn/pull/4/commits/108a83c30ff191a6a93b55c07e0bf1e1bb971ba1)

**Comment by tl240** [comment]
This PR is moved to [PR](https://github.com/icetea-labs/taskfi-tlsn/pull/1). Further discussion should be continued there.

**Pull request closed without merging**